File: src/main/java/info/rsdev/playlists/services/InternetChartsFetcher.java
package info.rsdev.playlists.services;

import java.io.IOException;
import java.util.Optional;

import org.apache.http.HttpStatus;
import org.jsoup.HttpStatusException;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.rsdev.playlists.exception.FailedHostException;

public record InternetChartsFetcher(String location) implements DocumentFetcher {

    private static final Logger LOGGER = LoggerFactory.getLogger(InternetChartsFetcher.class);

    private static final String ABORT_MSG = "Website down? Abort scraping to prevent holes in the chart data.";
    
    private static final String USERAGENT_STRING = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/117.0";

    @Override
    public Optional<Document> fetch() {
        long start = System.currentTimeMillis();
        try {
            return Optional.ofNullable(Jsoup.connect(location()).userAgent(USERAGENT_STRING).get());
        } catch (HttpStatusException e) {
            if (e.getStatusCode() == HttpStatus.SC_NOT_FOUND) {
                // we might query a non-existent week, e.g. week 53
                LOGGER.error("Could not fetch data from {}: {}", location(), e.getMessage());
                return Optional.empty();
            }
            throw new FailedHostException(ABORT_MSG, e);
        } catch (IOException e) {
            throw new FailedHostException(ABORT_MSG, e);
        } finally {
            if (LOGGER.isDebugEnabled()) {
                LOGGER.debug("Fetched {} in {} ms.", location(), System.currentTimeMillis() - start);
            }
        }
    }
}

File: src/main/kotlin/info/rsdev/playlists/services/InternetChartsFetcher.kt
/*
 * Copyright 2018 Red Star Development.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package info.rsdev.playlists.services

import org.jsoup.HttpStatusException
import org.jsoup.Jsoup
import org.jsoup.nodes.Document

private const val USERAGENT_STRING = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:59.0) Gecko/20100101 Firefox/59.0"

open class InternetChartsFetcher(private val location: String) : DocumentFetcher {

    override fun fetch(): Document? = try {
            Jsoup.connect(getLocation())
                .userAgent(USERAGENT_STRING)
                .get()
        } catch( e: HttpStatusException) { null }

    override fun getLocation() = this.location
}
